gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/MysqlUserQuotaManagerTest.java

```diff
@@ -0,0 +1,137 @@
+    this.quotaManager.decrementJobCount(TEST_NAME, AbstractUserQuotaManager.CountType.USER_COUNT);
+    Assert.assertEquals(this.quotaManager.getCount(TEST_NAME, AbstractUserQuotaManager.CountType.USER_COUNT), 1);
+
+    this.quotaManager.decrementJobCount(TEST_NAME, AbstractUserQuotaManager.CountType.USER_COUNT);
+    Assert.assertEquals(this.quotaManager.getCount(TEST_NAME, AbstractUserQuotaManager.CountType.USER_COUNT), 0);
+
+    this.quotaManager.decrementJobCount(TEST_NAME, AbstractUserQuotaManager.CountType.USER_COUNT);
+    Assert.assertEquals(this.quotaManager.getCount(TEST_NAME, AbstractUserQuotaManager.CountType.USER_COUNT), 0);
+
+    this.quotaManager.decrementJobCount(TEST_NAME, AbstractUserQuotaManager.CountType.FLOWGROUP_COUNT);
+    Assert.assertEquals(this.quotaManager.getCount(TEST_NAME, AbstractUserQuotaManager.CountType.FLOWGROUP_COUNT), 1);
+    this.quotaManager.decrementJobCount(TEST_NAME, AbstractUserQuotaManager.CountType.FLOWGROUP_COUNT);
+    Assert.assertEquals(this.quotaManager.getCount(TEST_NAME, AbstractUserQuotaManager.CountType.FLOWGROUP_COUNT), -1);
+  }
+
+  class ChangeCountRunnable implements Runnable {
+    boolean increaseOrDecrease;
+
+    public ChangeCountRunnable(boolean increaseOrDecrease) {
+      this.increaseOrDecrease = increaseOrDecrease;
+    }
+
+    @Override
+    public void run() {
+      int i = 0;
+      while (i++ < 1000) {
+        try {
+          if (increaseOrDecrease) {
+            MysqlUserQuotaManagerTest.this.quotaManager.incrementJobCount(TEST_NAME, AbstractUserQuotaManager.CountType.USER_COUNT);
+          } else {
+            MysqlUserQuotaManagerTest.this.quotaManager.decrementJobCount(TEST_NAME, AbstractUserQuotaManager.CountType.USER_COUNT);
+          }
+        } catch (IOException e) {
+          Assert.fail("Thread got an exception.", e);
+        }
+      }
+    }
+  }
+
+  @Test(dependsOnMethods = "testDecreaseCount")
+  public void testConcurrentChanges() throws IOException, InterruptedException {
+    Runnable increaseCountRunnable = new ChangeCountRunnable(true);
+    Runnable decreaseCountRunnable = new ChangeCountRunnable(false);
+    Thread thread1 = new Thread(increaseCountRunnable);
+    Thread thread2 = new Thread(increaseCountRunnable);
+    Thread thread3 = new Thread(increaseCountRunnable);
+    Thread thread4 = new Thread(decreaseCountRunnable);
+    Thread thread5 = new Thread(decreaseCountRunnable);
+    Thread thread6 = new Thread(decreaseCountRunnable);
+
+    thread1.start();
+    thread2.start();
+    thread3.start();
+    thread1.join();
+    thread2.join();
+    thread3.join();
+    Assert.assertEquals(this.quotaManager.getCount(TEST_NAME, AbstractUserQuotaManager.CountType.USER_COUNT), 3000);
+    thread4.start();
+    thread5.start();
+    thread6.start();
+    thread4.join();
+    thread5.join();
+    thread6.join();
+    Assert.assertEquals(this.quotaManager.getCount(TEST_NAME, AbstractUserQuotaManager.CountType.USER_COUNT), -1);
```

Comment:
why does this become -1, should it decrease by 3?

Reply:
Each of the three threads decrease it by 1000 times. so it the count should be reduced to 0. When count is 0, it also removes the row from the table and then a `get` returns -1